src/il_representations/algos/pair_constructors.py

```diff
@@ -167,3 +170,81 @@ def __call__(self, data_iter):
                 act_queue.reset()
             else:
                 timestep += 1
+
+
+class JigsawPairConstructor(TargetPairConstructor):
+    def __init__(self, permutation_path, n_tiles=9):
+        self.permutation_path = permutation_path
+        self.permutation = np.load(self.permutation_path)
+        self.n_tiles = n_tiles
+
+    def __call__(self, data_iter):
+        timestep = 0
+        traj_ind = 0
+
+        # Get tile dim
+        first_dict = next(data_iter)
+        original_h, original_w = first_dict['obs'][0].shape[-2], \
+                                 first_dict['obs'][0].shape[-1]
+
+        # We -2 after division here because, as the original paper indicated,
+        # preserving the whole image tile may let the network exploit edge information
+        # and not really learn useful representations. Therefore, here we are using
+        # the central-cropped version of these image tiles.
+        tile_h, tile_w = int(original_h / math.sqrt(self.n_tiles)) - 2, \
+                         int(original_w / math.sqrt(self.n_tiles)) - 2
+
+        permutation_class = [x for x in range(len(self.permutation))]
+        random.shuffle(permutation_class)
+        permute_idx = 0
+
+        for step_dict in data_iter:
+            # Process images into image tiles
+            obs = step_dict['obs']
+            processed_obs_list = []
+            target_list = []
+            for o in obs:
+                processed_obs = self.make_jigsaw_puzzle(o,
+                                                        permutation_class[permute_idx],
+                                                        tile_h,
+                                                        tile_w)
+                processed_obs_list.append(processed_obs)
+            target_list.append(permutation_class[permute_idx])
+            permute_idx = (permute_idx + 1) % len(self.permutation)
```

Comment:
Why do label assignment cyclically? You could just choose a random permutation class each time instead. That way there are no correlations between labels across batches.

```suggestion
            target_list.append(np.random.randint(len(self.permutation)))
```